src/main/java/org/magicdgs/thaplv/tools/ld/engine/QueueLD.java

```diff
@@ -0,0 +1,279 @@
+        return snpQueue.size();
+    }
+
+    /** Gets the number of computed pairs. */
+    public int computedPairs() {
+        return computedPairs.intValue();
+    }
+
+    /**
+     * Adds the variant to the queue and computes the queue if necessary.
+     */
+    public boolean add(final VariantContext variant) {
+        // get the contig
+        final String contig = variant.getContig();
+        if (currentContig == null) {
+            currentContig = contig;
+        }
+        // check if the queue is empty
+        if (snpQueue.isEmpty() ||
+                // or if it is n the contig and in the first variant, add directly
+                (currentContig.equals(contig) && binningParams.bellowMaximumDistance(
+                        variant.getStart(), snpQueue.peek().getPosition()))) {
+            // directly added
+            return snpQueue.add(new LightGenotype(variant));
+        }
+        // if it is not, check if it change the chromosome
+        if (snpQueue.peekLast().getContig().equals(contig)) {
+            // compute only for the firs variant
+            computeQueueLD();
+            // recursive call
+            return add(variant);
+        } else {
+            // empty the queue
+            while (!snpQueue.isEmpty()) {
+                computeQueueLD();
+            }
+            // write the bins
+            writeBins();
+            // we changed the contig
+            currentContig = contig;
+            // add variant to the queue
+            return snpQueue.add(new LightGenotype(variant));
+        }
+    }
+
+    /**
+     * Computes LD for the first element in the queue against all the other
+     */
+    private void computeQueueLD() {
+        // extract the first variant
+        final LightGenotype firstVariant = snpQueue.remove();
+        // TODO: filter by maf here?
+        for (final LightGenotype second : snpQueue) {
+            final LDcomputation toCompute = new LDcomputation(firstVariant, second);
+            // run the thread to compute LD in the ThreadPool
+            runner.accept(toCompute);
+        }
+    }
+
+    /**
+     * Writes the bins and reset them.
+     */
+    private void writeBins() {
+        // wait till completion
+        if (executor != null) {
+            waitUntilFinish();
+        }
+        output.write(currentContig);
+        output.clear();
+    }
+
+    /**
+     * Waits for the executor to finish all the loaded task.
+     */
+    private void waitUntilFinish() {
+        while (!executor.getQueue().isEmpty()) {
+            monitorLogging();
+        }
+        monitorLogging();
+    }
+
+    /**
+     * Logs the executor status with a monitor.
+     */
+    private synchronized void monitorLogging() {
+        if (executor != null) {
+            logger.debug(
+                    "[thread-pool] [{}/{}] Active: {}, Completed: {}, Task: {}, isShutdown: {}, isTerminated: {}",
+                    () -> new Object[] {
```

Comment:
is this working? probably add to the multi-thread test a DEBUG verbosity level. It could be too much, by the way...